feat(payment_providers): Add MoneyHash

shahwan42 opened the pull request:
## Roadmap Task

NA

## Feature Request
https://getlago.canny.io/feature-requests/p/integration-with-moneyhash-for-payment-orchestration

## Context

MoneyHash is a leading payment infrastructure software in Africa and the Middle East. Our biggest features are providing a superior payment orchestration for many payment providers in the region, smart payment routing among other features.

We have common customers, and it's a needed feature to have MoneyHash integrated into Lago.

This PR integrates MoneyHash as a payment provider for Lago's customers.

## Description

This PR implements the following changes
- When a new customer is created in Lago and the "sync with provider" checkbox is checked, a new customer is also created in MoneyHash.
- Create a payment URL for invoices, allowing customers to pay their invoices via MoneyHash
- Create Payment Requests that are processed using the customer's payment method.
- Webhook Handling
  - Transaction Webhooks
    - On the "transaction.purchase.successful" callback, the invoice/payment status is set to "Succeeded."
    - On the "transaction.purchase.failed" callback, the invoice/payment status is set to "Failed."
  - Intent Webhooks
    - On the "intent.time_expired" callback, the invoice/payment status is set to "Failed."
  - Card Token Webhooks
    - On the "card_token.created" and "card_token.updated" callbacks, the customer’s payment method ID is added or updated.
    - On the "card_token.deleted" callback, the customer’s payment method ID is removed.

List any dependencies that are required.
NA

julienbourdeau commented:
Hi @shahwan42,

This is awesome! Thank you contributing such a large feature.

I had a quick look and it looks great but first, we need to confirm with the product team how they feel about adding MoneyHash support.

We'll review it but I had a quick look and it looks great. 

We'll have to do some QA testing to feel comfortable maintaining it in the future. I'm sure you can help to get a testing account and for any question we'll have. 👍

shahwan42 commented:
Regarding this particular PR,
- no conflicts
- ran tests and they passed

Should I do anything else?
Thanks. @julienbourdeau

julienbourdeau commented on app/services/payment_providers/moneyhash/payments/create_service.rb:
```diff
@@ -0,0 +1,31 @@
+        end
+
+        def call
+          PaymentRequests::Payments::MoneyhashService.new(@invoice).create
```
I'd expect the opposite. The `PaymentRequests::Payments:: MoneyhashService` service to rely on the `PaymentProviders::Moneyhash::Payments::CreateService` to create the payment and then do specific things (like sending the email).

I think all the attributes are leftovers too.

julienbourdeau commented on app/services/payment_requests/payments/moneyhash_service.rb:
```diff
@@ -0,0 +1,222 @@
+        PaymentRequestMailer.with(payment_request: payment.payable).requested.deliver_later if result.payable.payment_failed?
+        result
+      rescue BaseService::FailedResult => e
+        result.fail_with_error!(e)
```
We recently fixed a bug where emails weren't sent if we couldn't create the payment, I think you want the same.
https://github.com/getlago/lago-api/pull/3155/files

julienbourdeau reviewed (commented):
Thank you! I had a first pass and left some initial comments 🙏

julienbourdeau commented:
Hey @shahwan42 ,

We merged a big style change, nothing important but replaced all single quotes by double quotes for strings. This created tons of conflict so you might have some difficulties rebase your PR.

I'd recommend doing it this way:
* squash all commits from your branch (we'll do this on merge any way)
* rebase on top of `main`
* to resolve all conflicts, take your changes: `git checkout --theirs .`
* continue rebase
* Then run `bundle exec rubocop -x --only Style/StringLiterals`
* commit

I can do it for you if you prefer!

shahwan42 commented on app/graphql/types/customers/object.rb:
```diff
@@ -126,16 +126,18 @@ def active_subscriptions_count
-        when :adyen
-          object.adyen_customer
-        end
+        return unless object
+
+        provider_mapping = {
+          stripe: :stripe_customer,
+          gocardless: :gocardless_customer,
+          cashfree: :cashfree_customer,
+          adyen: :adyen_customer,
+          moneyhash: :moneyhash_customer
+        }
+
+        method = provider_mapping[object.payment_provider&.to_sym]
+        object.public_send(method) if method
```
I did it because Rubocop complained because of the function complexity. Should I ignore it?

shahwan42 commented:
Hi @julienbourdeau
Thank you for the review, I handled your comments. Please take a look again once you can. Appreciated.

For the `Moneyhash::Payments::CreateService` part, since it's a different payment scenario, I've introduced new logic to handle the payment in this case. However, I'm not sure how to test it yet since we're only supporting automatic payments for subscription invoices only. Do you know how to make this scenario work by hand or using commands?

*There is some duplicate code since I have to move slowly and carefully because I'm still getting my hands into Ruby/Rails, and will improve it iteratively. Appreciate your effort and patience.*

julienbourdeau commented on app/services/payment_providers/moneyhash/customers/create_service.rb:
```diff
@@ -0,0 +1,79 @@
+        end
+
+        def generate_checkout_url(async)
+          return PaymentProviderCustomers::MoneyhashCheckoutUrlJob.perform_later(result.provider_customer) if async
```
I don't understand what happens here if it's async.

When it's synchronous, the checkout URL is generated and returned in the API response so you can redirect your customer. But if it's async, what do you do with URL? It doesn't seem saved anywhere.

julienbourdeau commented on app/controllers/webhooks_controller.rb:
```diff
@@ -74,4 +74,22 @@ def adyen
   end
+
+  def moneyhash
+    result = PaymentProviders::Moneyhash::HandleIncomingWebhookService.call(
+      organization_id: params[:organization_id],
+      code: params[:code].presence,
+      body: JSON.parse(request.body.read)
+    )
```
Instead of calling the service directly here, you should save the incoming webhook like we do for stripe. It will save the payload to the DB and only after dispatch jobs automatically. To process the event, you'll have to add your service to `InboundWebhooks::ProcessService::WEBHOOK_HANDLER_SERVICES`. 

This was added in December and only stripe was migrated yet so if you mainly looked at cashfree for instance, this isn't used yet. We'll migrate it.

```ruby
InboundWebhooks::CreateService.call(
      organization_id: params[:organization_id],
      webhook_source: :moneyhash,
      code: params[:code].presence,
      payload: request.body.read
    )
```

[/app/services/inbound_webhooks/process_service.rb#L5-L7](https://github.com/getlago/lago-api/blob/c6817cd5c9f0f9c3043b86fd137d84d7ed762e14/app/services/inbound_webhooks/process_service.rb#L5-L7)
```rb
    WEBHOOK_HANDLER_SERVICES = {
      stripe: PaymentProviders::Stripe::HandleIncomingWebhookService
    }
```

julienbourdeau commented:
Thank you @shahwan42 ! 

I left a few comments but I have 2 main concern now:
* Should we use PENDING or PROCESSING status. I'm trying to confirm which on is best so let me know why you went with pending 🙏 
* Tests 😬 

There are currently no tests. Ideally, we'd need test on:
* pricessing the webhooks (with fixtures payload)
* `app/services/invoices/payments/moneyhash_service.rb`
* `app/services/payment_requests/payments/moneyhash_service.rb`
* `app/services/payment_providers/moneyhash_service.rb`
* `app/services/payment_provider_customers/moneyhash_service.rb`

I'm assuming you're running the code locally and it works but we'd still need some basic tests mocking the API. I understand that if you are used to work with Rails it's not necessarily easy.

Can you rebase as well?

If you can also email me at julien@getlago.com I'd love to get a testing account so if we ever need to make some changes, we have access to the API.

shahwan42 commented:
Hi @julienbourdeau 

Thank you very much for your last review. I've gone through your comments and addressed them. Here are my replies to all your comments above.

**Payment Statuses**
I've moved the payment statuses to the moneyhash_provider file and the utility method `determine_payment_status` in the BaseProvider will be used to determine the payment_status and the `MoneyhashProvider#payable_payment_status`  will do the same for payables. I also commented how MoneyHash's statuses maps to Lago's payment statuses.

**CheckoutUrl**
When it's async, the checkout_url is sent to the merchant's webhook defined in Lago's dashboard. This is done by lago and was illustrated in the stripe payment video in the docs. I'm not sure if there's a way to save it in the db or how it's retrieved later, I haven't noticed it in other providers code.

**Webhook**
I updated webhook handling similar to Stripe, however I haven't implemented the signature validation yet, I'll delay it for a later update.

**Tests**
I've tested the services you mentioned, one by one.

**Misc**
During tests, I've found potential bugs/areas of improvements. Hence, I introduced some updates here and there.

**Demo**
I've uploaded a demo that illustrates how MH is supposed to be used (similar to Stripe's video from the official docs). Please review it if it helps. Link: https://jam.dev/c/70e79f6c-ea9a-4e77-bffc-3b7f7edc9b1e


**Re-review**
Please re-review and LMK if you've got any comments or concerns.

I appreciate your effort and patience. 🙏

julienbourdeau commented on app/services/payment_providers/moneyhash_service.rb:
```diff
@@ -0,0 +1,159 @@
+        )
+      end
+
+      moneyhash_provider.api_key
```
I believe here you had `old_key = moneyhash_provider.api_key` and since `old_key` wasn't used, rubocop remove the variable assignation. I think you can remove the line.

julienbourdeau commented on app/services/payment_providers/moneyhash_service.rb:
```diff
@@ -0,0 +1,159 @@
+      if payment_provider_code_changed?(moneyhash_provider, old_code, args)
+        moneyhash_provider.customers.update_all(payment_provider_code: args[:code]) # rubocop:disable Rails/SkipsModelValidations
+      end
+
```
Make API call to get the signature key from moneyhash. You can make it synchronous so we can return an error if something goes wrong or in a job if you prefer.
You can save the key in the payment_provider settings 👍

julienbourdeau commented:
Thanks for adding tests and addressing concerns. The fixtures are very useful as well!

I see 2 issues before we merge:
* Make sure you address this https://github.com/getlago/lago-api/pull/3117/files#r1977611301
* Implement signature validation

For the signature key, I'd make an call to moneyhash in `app/services/payment_providers/moneyhash_service.rb`, I added a comment.

Make sure you rebase and after that I believe we can merge! 🥳 

I'm very aware that adding a payment provider is a lot of work, thank you for your effort!

julienbourdeau commented:
@shahwan42  I'm happy to rebase your PR if needed but can you allow edits from maintainers?
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

shahwan42 commented:
Thank you, @julienbourdeau

I've done the signature verification, similar to Stripe's implementation with a key difference, it's synchronous and won't create the connection without getting the signature_key from MH's side. Please tell me if you got any concerns or notes.

For allowing the PR edits, it seems not possible since I've opened MH/main against Lago/main, and I don't have a PR open on MH's fork repo. If it's necessary/recommended, I could close this PR and open another one to allow maintainer edits. Please lmk.

Appreciate your review.

shahwan42 commented:
To save time, I'll close this PR and open it again to allow you to edit on it. Thanks.

shahwan42 commented:
Re-opened here: https://github.com/getlago/lago-api/pull/3409

shahwan42 commented:
It seems I don't have the permission to do that. I'll ask someone.
